exopy_hqc_legacy/tasks/tasks/util/fit_resonance_tasks.py

```diff
@@ -0,0 +1,467 @@
+    column_name_maglin = Unicode().tag(pref=True)
+    column_name_phase = Unicode().tag(pref=True)
+    column_name_freq = Unicode().tag(pref=True)
+
+    #: Flag indicating the measurement setup. This changes the fit function
+    mode = Enum('Reflection', 'Transmission', 'Lorentzian').tag(pref=True)
+
+    array_format = ['VNA']
+
+    def perform(self):
+        """ Find resonance frequency of database array and store index/value
+            pairs.
+
+        """
+        task = self.task
+        use_ref = task.use_ref
+        array = task.format_and_eval_string(task.target_array)
+        if use_ref:
+            array_ref = task.format_and_eval_string(task.ref_array)
+
+        freq = array[self.column_name_freq]
+        data_maglin = array[self.column_name_maglin]
+        delay = 111
+
+        if self.mode == 'Reflection':
+            data_phase = array[self.column_name_phase]
+            data_c = data_maglin*np.exp(1j*np.pi/180*data_phase)*np.exp(1j*2*np.pi*freq*delay*1e-9)
+            if use_ref:
+                freq_ref = array_ref[self.column_name_freq]
+                data_maglin_ref = array_ref[self.column_name_maglin]
+                data_phase_ref = array_ref[self.column_name_phase]
+                freq_ref, data_maglin_ref, data_phase_ref = adapt_ref(freq, freq_ref, data_maglin_ref, data_phase_ref)
+                data_c_ref = data_maglin_ref*np.exp(1j*np.pi/180*data_phase_ref)
+                data_c = data_c/data_c_ref
+
+        if self.mode == 'Lorentzian':
+            data_error = array[self.column_name_phase]
+        if self.mode == 'Reflection':
+            try:
+                val, fit_err = fit_complex_a_out(freq, data_c)
+            except:
+                val = 1e9
+                fit_err = 100
+        if self.mode == 'Transmission':
+            try:
+                print(np.shape(freq), np.shape(data_maglin))
+                val, fit_err = fit_lorentzian(freq, data_maglin)
+            except:
+                val = 1e9
+                fit_err = 100
+        if self.mode == 'Lorentzian':
+            try:
+                val, fit_err = fit_lorentzian(freq, data_maglin, error=data_error)
+            except:
+                val = 1e9
+                fit_err = 100
+
+        task.write_in_database('res_value', val)
+        print('freq = '+str(np.round(val*1e-9, 5)*1e9) + ', error = '+str(np.round(fit_err, 2)))
+        task.write_in_database('fit_err', fit_err)
+        if fit_err > 1:
+            log = logging.getLogger(__name__)
+            msg = ('Fit resonance has abnormally high fit error,'
+                   'freq fit = {} GHz, relative error = {}')
+            log.warning(msg.format(round(val*1e-9, 3), round(fit_err, 2)))
+
+    def check(self, *args, **kwargs):
+        """ Check the target array can be found and has the right column.
+
+        """
+#        test, traceback = super(FitResonanceTask, self).check(*args, **kwargs)
```

Comment:
Is that really reasonable ?

Reply:
It does not seem very reasonable indeed, but I will let Zaki adjust this part